assembly/src/serde/mod.rs

```diff
@@ -1,265 +1,280 @@
+        target.write_str(self)
+    }
+}
+
+impl<T> Serializable for Vec<T>
+where
+    T: Serializable,
+{
+    fn write_into(&self, target: &mut ByteWriter) -> Result<(), SerializationError> {
+        target.write_len(self.len())?;
+        self.iter().try_for_each(|t| t.write_into(target))
+    }
+}
+
+impl<T> Serializable for Option<T>
+where
+    T: Serializable,
+{
+    fn write_into(&self, target: &mut ByteWriter) -> Result<(), SerializationError> {
+        match self {
+            Some(t) => {
+                target.write_bool(true);
+                t.write_into(target)
+            }
+            None => {
+                target.write_bool(false);
+                Ok(())
+            }
+        }
+    }
+}
+
+// BYTE READER
+// ================================================================================================
+
+/// Contains bytes for deserialization and current reading position
+pub struct ByteReader<'a> {
+    bytes: &'a [u8],
+    pos: usize,
+}
+
+impl<'a> ByteReader<'a> {
+    pub fn new(bytes: &'a [u8]) -> Self {
+        ByteReader { bytes, pos: 0 }
+    }
+
+    pub fn read_bool(&mut self) -> Result<bool, SerializationError> {
+        self.check_eor(1)?;
+        let result = self.bytes[self.pos];
+        self.pos += 1;
+        u8_to_bool(result)
+    }
+
+    pub fn read_u8(&mut self) -> Result<u8, SerializationError> {
+        self.check_eor(1)?;
+        let result = self.bytes[self.pos];
+        self.pos += 1;
+        Ok(result)
+    }
+
+    pub fn peek_u8(&self) -> Result<u8, SerializationError> {
+        self.check_eor(1)?;
+        let result = self.bytes[self.pos];
+        Ok(result)
+    }
+
+    pub fn read_u16(&mut self) -> Result<u16, SerializationError> {
+        self.check_eor(2)?;
+        let result = &self.bytes[self.pos..self.pos + 2];
+        self.pos += 2;
+        Ok(u16::from_le_bytes(
+            result.try_into().expect("u16 conversion failure"),
+        ))
+    }
+
+    pub fn read_u32(&mut self) -> Result<u32, SerializationError> {
+        self.check_eor(4)?;
+        let result = &self.bytes[self.pos..self.pos + 4];
+        self.pos += 4;
+        Ok(u32::from_le_bytes(
+            result.try_into().expect("u32 conversion failure"),
+        ))
+    }
+
+    pub fn read_u64(&mut self) -> Result<u64, SerializationError> {
+        self.check_eor(8)?;
+        let result = &self.bytes[self.pos..self.pos + 8];
+        self.pos += 8;
+        Ok(u64::from_le_bytes(
+            result.try_into().expect("u64 conversion failure"),
+        ))
+    }
+
+    pub fn read_len(&mut self) -> Result<usize, SerializationError> {
+        self.read_u16().map(|l| l as usize)
+    }
+
+    pub fn read_felt(&mut self) -> Result<Felt, SerializationError> {
+        let value = self.read_u64()?;
+        if value >= Felt::MODULUS {
+            Err(SerializationError::InvalidFieldElement)
+        } else {
+            Ok(Felt::new(value))
+        }
```

Comment:
Unrelated to this PR: This feels to me we are leaking some details of `Felt` here. Ideally, we should have a method in `Felt` that will take a `u64` and perform the necessary checks.

Reply:
Agreed! Though this would require a change in Winterfell which we won't be able to use for a while.